<b> Проект: SQL запросы. Анализ базы данных сервиса для чтения книг по подписке.</b>

<b> Оглавление проекта:</b>

[1. Загрузка и изучение данных.](#step_1)

[2. Расчет количества книг, вышедших после 1 января 2000 года.](#step_2)

[3. Расчет количества обзоров и среднюю оценку для кадой книги.](#step_3)

[4. Определение издательства, которое выпустило наибольшее число книг толще 50 страниц.](#step_4)

[5. Определение автора с самой высокой средней оценкой книг — (учитываются только книги с 50 и более оценками).](#step_5)

[6. Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок.](#step_6)
[7. Вывод.](#step_7)

<a id="step_1"></a> 
## Загрузка и изучение данных.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
def show_tabs(sql):
    return pd.io.sql.read_sql(sql, con = engine)

In [4]:
def show_info(df):
    return df.info()

In [5]:
query_1 = ''' SELECT * FROM books LIMIT 5 '''
books = show_tabs(query_1)
display(books)
books.info()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           5 non-null      int64 
 1   author_id         5 non-null      int64 
 2   title             5 non-null      object
 3   num_pages         5 non-null      int64 
 4   publication_date  5 non-null      object
 5   publisher_id      5 non-null      int64 
dtypes: int64(4), object(2)
memory usage: 368.0+ bytes


In [6]:
query_2 = ''' SELECT * FROM authors LIMIT 5 '''
authors = show_tabs(query_2)
display(authors)
authors.info()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  5 non-null      int64 
 1   author     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


In [7]:
query_3 = ''' SELECT * FROM publishers LIMIT 5 '''
publishers = show_tabs(query_3)
display(publishers)
publishers.info()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  5 non-null      int64 
 1   publisher     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


In [8]:
query_4 = ''' SELECT * FROM ratings LIMIT 5 '''
ratings = show_tabs(query_4)
display(ratings)
ratings.info()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   rating     5 non-null      int64 
dtypes: int64(3), object(1)
memory usage: 288.0+ bytes


In [9]:
query_5 = ''' SELECT * FROM reviews LIMIT 5 '''
reviews = show_tabs(query_5)
display(reviews)
reviews.info()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   text       5 non-null      object
dtypes: int64(2), object(2)
memory usage: 288.0+ bytes


<a id="step_2"></a> 
## Расчет количества книг, вышедших после 1 января 2000 года

In [10]:
query_count_books = ''' SELECT COUNT(book_id) FROM books WHERE CAST(publication_date AS date) >= '2000-01-01'; '''
count_books = show_tabs(query_count_books)
count_books

count
0    821

После 1 января 2000 года было опубликовано 821 книг.

<a id="step_3"></a> 
## Расчет количества обзоров и средней оценки для кадой книги

In [11]:
query_books_rat = '''
SELECT b.title,
       b.book_id,
       COUNT(DISTINCT rev.review_id) AS review_count,
       AVG(rat.rating) AS avg_rating
FROM books AS b 
LEFT JOIN ratings AS rat ON rat.book_id = b.book_id
LEFT JOIN reviews AS rev ON rev.book_id = b.book_id
GROUP BY b.book_id
ORDER BY avg_rating DESC , review_count DESC;
'''
books_rating = show_tabs(query_books_rat)
books_rating

title  book_id  review_count  \
0                        A Dirty Job (Grim Reaper  #1)       17             4   
1              School's Out—Forever (Maximum Ride  #2)      553             3   
2         Moneyball: The Art of Winning an Unfair Game      444             3   
3        Arrows of the Queen (Heralds of Valdemar  #1)       86             2   
4    Wherever You Go  There You Are: Mindfulness Me...      972             2   
..                                                 ...      ...           ...   
995  The World Is Flat: A Brief History of the Twen...      915             3   
996                                      Drowning Ruth      202             3   
997                  His Excellency: George Washington      316             2   
998                                              Junky      371             2   
999                               Harvesting the Heart      303             2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 4 columns]

Самый большой рейтинг и количество обзоров у книги `A Dirty Job (Grim Reaper #1)`

Наименьший рейтинг и количество обзоров у книги `Harvesting the Heart`

<a id="step_4"></a> 
## Определение издательства, которое выпустило наибольшее число книг толще 50 страниц

In [12]:
query_top_publisher = '''
WITH
id AS (SELECT publisher_id,
       COUNT(book_id) AS book_count
       FROM books
       WHERE num_pages > 50
       GROUP BY publisher_id
       ORDER BY book_count DESC
       LIMIT 1)
SELECT publisher,
       id.book_count
FROM publishers AS p RIGHT JOIN id ON p.publisher_id = id.publisher_id
'''
top_publisher = show_tabs(query_top_publisher)
top_publisher

publisher  book_count
0  Penguin Books          42

Издательство, которое выпустило наибольшее число книг, толще 50 страниц - `Penguin Books`, количество книг, выпущенное данным издательством составило 42 книги.

<a id="step_5"></a> 
## Определение автора с самой высокой средней оценкой книг — (учитываются только книги с 50 и более оценками)

In [16]:
query_top_author = '''
            SELECT a.author,
                   AVG(ratings)
            FROM (SELECT book_id,
                  AVG(rating) ratings
                  FROM ratings
                  GROUP BY book_id
                  HAVING COUNT(rating_id) >= 50) rat
            JOIN (SELECT book_id, author_id FROM books) b ON rat.book_id=b.book_id
            JOIN (SELECT author_id, author FROM authors) a ON b.author_id=a.author_id
            GROUP BY a.author
            ORDER BY avg(ratings) DESC
            LIMIT 1
        '''
top_author = show_tabs(query_top_author)
top_author

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

Автор с самой высокой средней оценкой книг — `J.K. Rowling/Mary GrandPré` , средняя оценка книг данного автора `4,28`.

<a id="step_6"></a> 
## Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
query_avg_reviews = '''
WITH
rat AS (SELECT *
        FROM (SELECT username,
              COUNT(rating_id) as number_of_ratings
              FROM ratings
              GROUP BY username) AS ref
        WHERE number_of_ratings > 50)
SELECT ROUND(AVG(count_reviews),2) 
FROM (SELECT COUNT(review_id) AS count_reviews
      FROM rat LEFT JOIN reviews AS rev ON rat.username = rev.username
      GROUP BY rev.username) AS ref_2
'''
reviews_count_1 = show_tabs(query_avg_reviews)
reviews_count_1

round
0  24.33

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет - `24,33`.

Был проведен анализ базы данных сервиса для чтения книг по подписке. Перечислим результаты проведенного анализа:

1. После 1 января 2000 года было опубликовано 821 книг.
2. Самый большой рейтинг и количество обзоров у книги `A Dirty Job (Grim Reaper #1)`
3. Наименьший рейтинг и количество обзоров у книги `Harvesting the Heart
4. Издательство, которое выпустило наибольшее число книг, толще 50 страниц - `Penguin Books`, количество книг, выпущенное данным издательством составило 42 книги.
5. Автор с самой высокой средней оценкой книг — `J.K. Rowling/Mary GrandPré` , средняя оценка книг данного автора `4,28`.
6. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет - `24,33`.